src/dataflow-types/src/client.rs

```diff
@@ -1019,123 +1024,115 @@ pub mod tcp {
                                     &self.addr,
                                     e,
                                     self.backoff.as_millis()
                                 );
-                                self.state =
-                                    GrpcComputedTcpConn::Backoff(Instant::now() + self.backoff);
+                                self.state = GrpcTcpConn::Backoff(Instant::now() + self.backoff);
                             }
                         }
                     }
-                    GrpcComputedTcpConn::AwaitResponse(client) => {
+                    GrpcTcpConn::AwaitResponse(client) => {
                         // The channel size is a arbitrary, but it should be a
                         // small bounded channel such that backpressure is applied early.
                         let (tx, rx) = mpsc::unbounded_channel();
-                        match client
-                            .command_response_stream(UnboundedReceiverStream::new(rx))
-                            .await
-                        {
-                            Ok(resp_rx_wrap) => {
-                                let resp_rx = resp_rx_wrap.into_inner();
-                                info!("GrpcComputedClient {}: connected", &self.addr);
-                                self.state = GrpcComputedTcpConn::Connected((tx, resp_rx));
+                        match client.create_stream(rx).await {
+                            Ok(stream) => {
+                                info!("GrpcClient {}: connected", &self.addr);
+                                self.state = GrpcTcpConn::Connected((tx, stream));
                             }
                             Err(err) => {
-                                debug!(
-                                    "GrpcComputedClient {}: Connection refused: {}",
-                                    &self.addr, err
-                                );
-                                self.state = GrpcComputedTcpConn::Disconnected;
+                                debug!("GrpcClient {}: Connection refused: {}", &self.addr, err);
+                                self.state = GrpcTcpConn::Disconnected;
                             }
                         }
                     }
-                    GrpcComputedTcpConn::Connected(_) => break,
-                    GrpcComputedTcpConn::Backoff(deadline) => {
+                    GrpcTcpConn::Connected(_) => break,
+                    GrpcTcpConn::Backoff(deadline) => {
                         time::sleep_until(*deadline).await;
-                        self.state = GrpcComputedTcpConn::Disconnected;
+                        self.state = GrpcTcpConn::Disconnected;
                     }
                 }
             }
         }
     }
 
     #[async_trait]
-    impl<T> GenericClient<ComputeCommand<T>, ComputeResponse<T>> for GrpcComputedClient
+    impl<Client, C, R, PC, PR> GenericClient<C, R> for GrpcClient<Client, PC, PR>
```

Comment:
We could think about making the command/response types associated types of the `GenericClient` trait, which might remove some generic parameters here and elsewhere.

Reply:
ooh that would be a nice refactor!